adding example calendar page (that works on its own)

vsoch opened the pull request:
@Chilipp here is the calendar code that should theoretically work (it works on a page on it's own, and also with the USRSE template). The error I get is https://stackoverflow.com/questions/26252938/fullcalendar-typeerror-fullcalendar-is-not-a-function. I'm going to work on the usrse calendar later today so minimally if this is able to work we know how to add events!

vsoch committed:
adding example calendar page (that works on its own)

Signed-off-by: vsoch <vsochat@stanford.edu>

Chilipp commented:
@vsoch, is there a reason why you choose such an outdated version of fullcalendar.io? I could get it rendered with with the latest version of fullcalendar, although the `dayGridMonth` does not work for whatever reason (it does not even produce an error...). see the [update-fullcalendar](https://github.com/SORSE/sorse.github.io/tree/update-fullcalendar) and particularly https://431-267395254-gh.circle-artifacts.com/0/SORSE/programme/calendar/index.html

`listweek` and `listday` works well, `dayGridMonth` does not...

vsoch commented:
I tried the most up to date first, and it also rendered a completely blank calendar, so I fell back to a getting  started (proven working) example. It feels like a fairly heavy library, I almost got it fully working for how I'd want it for USRSE but ran into needing a plugin to do an "Nth day of the month" sort of deal, so I stopped there. When things get too complicated like that, or can easily break, I tend to step back and say "is this really necessary or can it be done in a simpler way?"

Chilipp commented:
I see. I'd like to give it another try tomorrow, but yes, if it turns out to be too complicated we can also go with a simpler jekyll-like solution.

Chilipp committed:
Update fullcalendar (#213)

* use latest version of fullcalendar
and add a demo calendar
* override fc-daygrid display property
minimal-mistakes sets display: block for all tables which causes problem when displaying fullcalendar
* add some styles for the calendar

vsoch committed:
adding basic events feed to calendar

Signed-off-by: vsoch <vsochat@stanford.edu>

vsoch commented:
Okay I just rendered a very basic feed of the events, so they show up and the user can click to go to the event page. For events that have times (or more generally extend longer than one day) I think an `end_date` might be added.

Chilipp commented on _pages/programme/calendar.md:
```diff
@@ -0,0 +1,43 @@
+     events: [{% for event in site.events %}{
+       title  : '{{ event.title }}',
+       url    : '{{ event.url }}',
+       allDay : {% if event.all_day %}true{% else %}false{% endif %},{% if event.end_date %}end    : '{{ event.end_date | date_to_xmlschema }}'{% endif %}
```
```suggestion
       allDay : {% if event.all_day %}true{% else %}false{% endif %},
       {% if event.end_date %}end    : '{{ event.end_date | date_to_xmlschema }}'{% endif %}
```
just for readability

vsoch replied:
Are you sure? That will render an empty / blank line for most entries.

Chilipp replied:
hmm, to me the readability of the code is more important than how it renders. But if you like, we can also do

```suggestion
       allDay : {% if event.all_day %}true{% else %}false{% endif %},
       {% if event.end_date %}end    : '{{ event.end_date | date_to_xmlschema }}'{% else %}//{% endif %}
```
which puts two `//` instead of a blank line

Or maybe even

```suggestion
       allDay : {% if event.all_day %}true{% else %}false{% endif %},
       {% unless event.end_date %}// {% endunless %}end    : '{{ event.end_date | date_to_xmlschema }}'
```

Chilipp commented:
Looks good to me @vsoch! We should also add the rolling deadlines to the calendar (there, copy-paste is probably the easiest as it's always the last day of the month) and we need to think about the coloring (see SORSE/sorse.github.io#212), but we can talk about this in a follow-up PR.

I'd add this calendar actually at the top of the [/programme/ page](http://www.philipp-s-sommer.de/sorse.github.io/programme/), what do you think?

vsoch commented:
Hmm, for the top of the programme page I was thinking a list of upcoming events might be more appropriate, and then have a link to the calendar. I'm worried it takes up a lot of screen space and we'd want people to see that top filter of event type, and the events below, first thing. What do you think?

vsoch commented:
Something like https://fullcalendar.io/docs/list-view ?

Chilipp commented:
Hmm ok. We definitely need the upcoming events on the [front page](http://www.philipp-s-sommer.de/sorse.github.io/#upcoming-events), so this would be a duplication, wouldn't it? But I see your point, the `dayGridMonth` really takes a lot of space so it's probably better to keep it separate.

I also thought about a list-view for the upcoming events. Is there a way to only shows future events?

vsoch commented:
Yes! You would want to compare the event date with the current jekyll site date (`{{ site.time }} maybe?) and then check if it's later / earlier. Here is an example of how we did it with an expires data: https://raw.githubusercontent.com/USRSE/usrse.github.io/master/pages/resources/events-training.md. So something like:

```
{% assign sorted = (site.events | sort: 'date') %}
{% for event in sorted %}{% capture nowunix %}{{'now' | date: '%s'}}{% endcapture %}{% capture event_date %}{{ event.date | date: '%s'}}{% endcapture %}{% if event_date > nowunix %}
## This is a future event!
{% endif %}{% endfor %}
```
I didn't test that (just threw it together) but I think something along those lines would work.

Chilipp commented:
Sounds good! Then we should combine this with an automated daily rebuild of the website, shouldn't we? Otherwise we could also do it with pure Javascript, e.g.

```js
var events = [
    {% for event in site.events %}
    {
       title  : '{{ event.title }}',
       url    : '{{ event.url }}',
       allDay : {% if event.all_day %}true{% else %}false{% endif %},{% if event.end_date %}end    : '{{ event.end_date | date_to_xmlschema }}'{% endif %}
       start  : '{{ event.date | date_to_xmlschema }}'
    }{% if forloop.last %}{% else %},{% endif %}
    {% endfor %}
];
events = events.filter(e => new Date(e.start) >= new Date());
events.sort((e1, e2) => new Date(e1.start) - new Date(e2.start));
events = events.slice(0, 10); // to only show the first 10
var calendar = new FullCalendar.Calendar(calendarEl, {
 initialView: 'listYear',
 ....
 events: events
});
```

vsoch commented:
Yes the javascript approach might be better suited - the example I derived from was using an expired date, which wouldn't be a moving target like a current time :)

Chilipp commented:
alright, shall we resolve the above conversation, merge this PR and continue in a new one for the upcoming events (as this one was about the calendar page)?

It's already pretty late over here, so I won't be able to do more today.

vsoch commented:
Sure thing, works for me :)

Chilipp committed:
include line break in rendered JSON
